### PR TITLE
Handle realm close in did_change callbacks

### PR DESCRIPTION
**Pull request by beeender**
Bindings could close Realm in the notify callbacks.

### CODE REVIEW

File: tests/realm.cpp [hunk 1]
```diff
@@ -383,3 +383,95 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
...
+
+            virtual void did_change(std::vector<ObserverState> const&, std::vector<void*> const&, bool) override
+            {
+                REQUIRE(false);
+            }
+        };
+        r1->m_binding_context.reset(new Context(r1));
+    }
+
+    SECTION("will_change") {
```
Comment (beeender) [inline tests/realm.cpp]: @tgoyne Are `will_change` and `did_change` still needed after your object notification work? Shall I remove the `will_change` test from this PR? It actually requires more modifications. And I don't think Realm instance should be closed in the `will_change` callback.

File: src/shared_realm.cpp [hunk 2]
```diff
@@ -575,17 +575,23 @@ void Realm::notify()
     m_is_sending_notifications = true;
     auto cleanup = util::make_scope_exit([this]() noexcept { m_is_sending_notifications = false; });
 
+    // The SharedRealm could be closed in the callbacks. Keep a ref locally to check if Realm gets closed.
+    auto this_ref = shared_from_this();
     if (m_shared_group->has_changed()) { // Throws
         if (m_binding_context) {
             m_binding_context->changes_available();
+            if (this_ref->is_closed()) return;
```
Comment (tgoyne) [inline src/shared_realm.cpp]: Newline before `return` for all of these.

File: tests/realm.cpp (hunk 1 above)
Comment (tgoyne) [inline tests/realm.cpp]: Yes, will_change() and did_change() are still needed and their API and semantics are unchanged. Not handling the Realm being closed from within them is perfectly acceptable.

File: tests/realm.cpp [hunk 3]
```diff
@@ -383,3 +383,95 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
...
+    SECTION("changes_available") {
+        struct Context : public TestContext {
+            Context(SharedRealm& r) : TestContext(r) { }
+            virtual void changes_available() override
+            {
+                shared_realm->close();
+                shared_realm.reset();
+            }
+
+            virtual void will_change(std::vector<ObserverState> const&, std::vector<void*> const&) override
```
Comment (tgoyne) [inline tests/realm.cpp]: `virtual` is redundant when the function is marked `override`.

File: src/shared_realm.cpp (hunk 2 above)
Comment (tgoyne) [inline src/shared_realm.cpp]: Calling these via `this_ref->` suggests that it's somehow different from `this`, when it's actually fine to just call `is_closed()` directly.

**Comment by beeender** [comment]
@tgoyne java is using `did_change()` for Realm notifications, so it would be quite important to handle the close case with all `did_change()` called. The problem is this requires to change 
```cpp
    TransactLogObserver(BindingContext* context, SharedGroup& sg, Func&& func,
                        _impl::NotifierPackage& notifiers, bool validate=true)
```
to take a `SharedRealm` instead of `SharedGroup`. Otherwise the sg might become an invalid reference when the Realm gets closed. Is it an acceptable change or there are some other ideas to handle this case?

**Comment by tgoyne** [comment]
The transaction log functions are used in places where the SharedGroup is not owned by a Realm. An awkward option would be to instead pass in a const ref to the `unique_ptr<SharedGroup>` and check that that's non-null all over the place.

**Comment by beeender** [comment]
@tgoyne PR updated
1. Java only exposes `did_change()` and collection callback to end users. So to make this PR simpler, it only handles the `did_change()`.
2. Will it be a problem if user holding multiple collection notification tokens and close the Realm in the first collection change callback?
3. The test case `table change information / modifications to untracked tables are ignored` fails, but I cannot figure out why. It doesn't seem that any changes could cause that? Would you please kindly help to take a look? thanks.

File: src/impl/transact_log_handler.cpp [hunk 4]
```diff
@@ -657,20 +657,23 @@ void advance_with_notifications(BindingContext* context, SharedGroup& sg, Func&&
     if (observers.empty() && (!notifiers || notifiers.version())) {
         notifiers.before_advance();
         func(TransactLogValidator());
-        auto new_version = sg.get_version_of_current_transaction();
+        auto new_version = sg->get_version_of_current_transaction();
         if (context && old_version != new_version)
             context->did_change({}, {});
+        // did_change() could close the Realm. Just return if it does.
+        if (!sg)
```
Comment (beeender) [inline src/impl/transact_log_handler.cpp]: Is it fine to return here?

File: src/shared_realm.cpp [hunk 5]
```diff
@@ -607,10 +607,14 @@ void Realm::notify()
             m_coordinator->advance_to_ready(*this);
         }
         else  {
+            // In case the SharedRealm gets descturcted in the did_change callback.
+            auto realm_ref = this->shared_from_this();
```
Comment (tgoyne) [inline src/shared_realm.cpp]: This doesn't actually work, because `cleanup`'s destructor will run after this reference is released and that needs `this` to still exist. The immediate fix would be to move this to the top of the function, but I'm not sure why the `Realm` instance would be destroyed from within the callback. There's no real need for the Java `Realm` to release its reference to the OS `Realm` after calling `close()` since it's an insignificant amount of memory freed up by destroying it early.

Also "descturcted", and the SharedGroup isn't what matters here.

File: tests/realm.cpp [hunk 6]
```diff
@@ -471,3 +474,67 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
...
+        r1->notify();
+    }
+
+    SECTION("did_change with async results") {
+        r1->m_binding_context.reset(new Context());
+        Results results(r1, table->where());
+        auto token = results.add_notification_callback([&](CollectionChangeSet, std::exception_ptr) {
+                // Should not be called.
+                REQUIRE(false);
+                });
```
Comment (tgoyne) [inline tests/realm.cpp]: Indentation is wrong.

File: tests/realm.cpp [hunk 7]
```diff
@@ -471,3 +474,67 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
...
+    r1->begin_transaction();
+    auto table = r1->read_group().get_table("class_object");
+    auto row_idx = table->add_empty_row(1);
+    auto table_idx = table->get_index_in_group();
+    r1->commit_transaction();
+
+    // Cannot be a member var of Context since Realm.close will free the context.
+    static SharedRealm* shared_realm;
+    shared_realm = &r1;
+    struct Context : public BindingContext {
```
Comment (tgoyne) [inline tests/realm.cpp]: Context can just hold a pointer to a stack variable rather than making it a static.

File: tests/realm.cpp [hunk 8]
```diff
@@ -471,3 +474,67 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
...
+        void did_change(std::vector<ObserverState> const&, std::vector<void*> const&, bool) override
+        {
+            (*shared_realm)->close();
+            (*shared_realm).reset();
+        }
+    };
+
+    SECTION("did_change") {
+        r1->m_binding_context.reset(new Context());
+        r1->invalidate();
```
Comment (tgoyne) [inline tests/realm.cpp]: This doesn't seem relevant to the test.

File: tests/transaction_log_parsing.cpp [hunk 9]
```diff
@@ -1325,8 +1325,8 @@ TEST_CASE("Transaction log parsing: changeset calcuation") {
 
         auto observe = [&](std::initializer_list<Row> rows, auto&& fn) {
             auto history = make_in_realm_history(config.path);
-            SharedGroup sg(*history, config.options());
-            auto& group = sg.begin_read();
+            std::unique_ptr<SharedGroup> sg = std::make_unique<SharedGroup>(*history, config.options());
```
Comment (tgoyne) [inline tests/transaction_log_parsing.cpp]: This is a good place to use `auto` as otherwise the type of the variable is specified twice.

File: src/impl/transact_log_handler.cpp (hunk 4 above)
Comment (tgoyne) [inline src/impl/transact_log_handler.cpp]: Yes, destroying the `NotifierPackage` without delivering it should be fine and not delivering the notifications seems like the only sensible thing to do.

File: src/impl/realm_coordinator.cpp [hunk 10]
```diff
@@ -330,16 +330,19 @@ void RealmCoordinator::commit_write(Realm& realm)
...
         bool have_notifiers = std::any_of(m_notifiers.begin(), m_notifiers.end(),
                                           [&](auto&& notifier) { return notifier->is_for_realm(realm); });
         if (have_notifiers) {
-            m_notifier_skip_version = Realm::Internal::get_shared_group(realm).get_version_of_current_transaction();
+            m_notifier_skip_version = Realm::Internal::get_shared_group(realm)->get_version_of_current_transaction();
         }
     }
 
+    // In case the SharedRealm gets descturcted in the did_change callback.
+    auto realm_ref = realm.shared_from_this();
```
Comment (tgoyne) [inline src/impl/realm_coordinator.cpp]: My comment about this in shared_realm.cpp applies to here as well (`realm` can theoretically be the thing keeping the coordinator alive).

File: tests/realm.cpp (hunk 7 above)
Comment (beeender) [inline tests/realm.cpp]: yes, the problem is when close the Realm, the context object will be freed. the `Context.m_shared_realm_ptr` itself become invalid.

File: src/shared_realm.cpp (hunk 5 above)
Comment (beeender) [inline src/shared_realm.cpp]: hmmm, that is a good point. Let me check if it is doable to only call close in java `Realm.close()` and delete the `SharedRealm` when it gets GCed.

Comment (beeender) [inline src/shared_realm.cpp]: I think there is no way to avoid java binding free the pointer in the callback because of the GC is not predictable. Like below example:

```java
// Member var
Realm realm;
void UserDefinedCallback() {
    realm.close(); // Here is totally doable that java keeps the SharedRealm alive.
    realm = null; // But we cannot really control what will happen here. The realm instance could be GCed at any time, which means the pointer it holds could be freed at any time from a different thread.
}
```

I use the immediate fix to move the `shared_from_this()` to the top of the method body.

File: tests/realm.cpp (hunk 8 above)
Comment (beeender) [inline tests/realm.cpp]: To make the `m_group` null so it can run with the below `else` routine in `notify()`:

```c++
    if (m_auto_refresh) {
        if (m_group) {
            m_coordinator->advance_to_ready(*this);
        }
        else  {
            if (m_binding_context) {
                m_binding_context->did_change({}, {});
            }
            if (!is_closed()) {
                m_coordinator->process_available_async(*this);
            }
        }
    }

```

**Comment by beeender** [comment]
@tgoyne updated, please help to review again. thanks

File: src/impl/realm_coordinator.cpp [hunk 11]
```diff
@@ -348,9 +351,9 @@ void RealmCoordinator::commit_write(Realm& realm)
         m_notifier->notify_others();
     }
 #if REALM_ENABLE_SYNC
-    if (m_sync_session) {
+    if (m_sync_session && !realm.is_closed()) {
```
Comment (tgoyne) [inline src/impl/realm_coordinator.cpp]: Shouldn't this be moved to before the notifications? I wouldn't expect calling `close()` on a Realm to stop it from syncing the commit I just made.

Comment (beeender) [inline src/impl/realm_coordinator.cpp]: Yup, you are right . Moved. @austinzheng would you please take a look of my last commit?
